### PR TITLE
The Value of Automated e2e-testing in a software project

### PR DESCRIPTION
**Pull request by riikkanen**

### CODE REVIEW

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+- Wicket
+- Selenium
+- programming
+---
+
+The software project concerned started in the beginning of Autumn in 2010. I had the honor to work in the project from its beginning to June 2014 when my maternity leave began. The project was about to renew a 30 years old, technically outdated registry and everything attached to it, such as the user inteface and all the integrations. The project was a success, after almost three years of hard work, the application was taken into serious use without any major hardships. The project is still alive and kicking, even though, nowadays mostly maintenance, but also some feature development. 
+
+##The application
+
+The application is coded with Java. The actual user interface of the registry is developed as a web application using Apache Wicket, which is an Java web application framework. The data is stored in Oracle database and queried with direct SQL-queries, without any intermediate layer such as Hibernate. There are also numerous integrations bubbling under, powered by Mule. In the project I worked as a software developer both in back and front ends.
```
Comment (juhofriman) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "coded with Java" kuulostaa jotenkin hassulta ja finglishmiltä. The application is made with Java? The application is written in Java? En tiedä!
Voisko viimeinen lause olla myös jotain: "I worked as a full stack developer"? Sen saisi ehkä jotenkin ton ekan lauseen kanssa myös yhteen luontevasti, niin lähtee kappale käyntiin siististi :)

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+At least for me, the many ways to show data were also a burden. The functionality behind hiding and showing the icons respectively was very easy to break by sincere mistake when implementing some new feature or tweaking an existing one. Also, the number of different permutations were so many that it was nearly impossible to make full regression testing just by clicking the web site manually. I think that was the main reason we started to write automated end-to-end-tests in quite an early phase of the project.
```
Comment (juhofriman) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: ..the many ways of showing the data..?
the number of different permutations were so vast that it was nearly ..?
I tend to think this was the... || I feel this was the ?

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+![An example of JBehave and Selenium in action](/img/value-of-e2e-testing/jbehave.png)
+
+At first, there was a huge gap between code implemented and code tested with JBehave tests. There were a couple of guys just writing tests for the already implemented features. However, soon it was clear, that the gap would never be surpassed this way. Everyone in the project should learn writing the new type of tests. In my opinion, in the beginning it was slow and not very easy but bit by bit it became a natural way to write tests. Of course, the test framework developed side by side with the tests, because we wrote a lot of common utility methods and tools.
```
Comment (juhofriman) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: guys? All male testing panel? :) Tuo on varmasti ihan totta, mutta voisi löytyä joku muu ilmaus?
bit by bit -> step by step it?
...the test framework was developed (?)...

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+The tests stories can, in principle, be used in communication between customers and developers or as a basis of user manual. However, soon we found out that keeping the stories in such a documentative way, is not easy and also not worth it. Also, the lack of IDE support for writing story files and matching the steps in the stories and in the test code is a little frustrating. It would be nice to find mismatches because of typos before you run the test, or navigate directly from a story step to the code implementing the step. One can say, that a number of text searches were done to avoid the problem. Furthermore, the finding right web elements to assert is not effortless with xpath either.
```
Comment (juhofriman) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: To some extent tests stories can be used in... (?)
is not easy and also not worth it. -> is hard and on top of that not really beneficial. (?)

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+##The Value Obtained
+
+The comprehensive regression test suite including both the e2e tests and the unit tests gives depelopes courage to refactor the code base more freely. After all, all development is easier, because the mistakes are quite likely to expose on some layer of tests. This is extremely valuable when new developers are introduced to the project. In a long-lasting project you cannot avoid coders changing, some are coming and some are going. Of course, the rookies cannot have the all functionality and information in their heads as the experienced, tough warriors. Hence, the automated tests lower the step when entering to a new project, and as I think, ease the mental burden of both newcomers and old-timers. I would say, that extensive testing also enables healthy and needed rotation of developers between projects. The extensive automated testing is crucial also in the maintenance phase of project. The changes to code base might be done less frequently and the developers making the changes may not be the original coders any more.
```
Comment (juhofriman) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: coders -> programmers? developers? personnel?

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
+---
+layout: post
+author: riikkanen
+title: >
+  The Value of Automated e2e-testing in a Software Project
+excerpt: >
+  A peek into automated e2e-testing done in a software project. The first try with Wicket Page Tester. The second and more sophisticated version with JBehave and Selenium.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "A peek into automated e2e-testing done in a software project." -> "done" can be left out.

Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: Would: "with Wicket Page Tester and JBehave/Selenium" suffice? With or without respectively.

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+- Java
+- testing
+- e2e testing
+- JBehave
+- Wicket
+- Selenium
+- programming
+---
+
+The software project concerned started in the beginning of Autumn in 2010. I had the honor to work in the project as a full stack developer from its beginning to June 2014 when my maternity leave began. The project was about to renew a 30 years old, technically outdated registry and everything attached to it, such as the user inteface and all the integrations. The project was a success, after almost three years of hard work, the application was taken into serious use without any major hardships. The project is still alive and kicking, even though, nowadays mostly maintenance, but also some feature development.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: Autumn in lowercase, drop the in between autumn and 2010.

"The project was about to renew" -> The project's goal was?

"The project was a success, after almost three years of hard work, the application was taken into serious use without any major hardships." -> After three years of hard work the application was successfully taken into use in production?

"The project is still alive and kicking, even though, nowadays mostly maintenance, but also some feature development. " -> Lots of commas: "The project is still alive and kicking in maintenance mode with some small feature development"?

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+The views in the user interface are designed in such a way that the user can see the main item and all the sub-items related at the same time. The main item could have several sub-items listed in a table-like manner. The main item has certain managing tools and actions and, of course, all the sub-items have a bunch of tools and actions too. Most of the tools are implemented as icons with actions attached. The state and visibility of the tools depend about the status of the main item, but also from the context. Is this the main item which created the sub-item, the main item which edited them or a main item having nothing to do with the others? In addition to, some notification icons need to be shown in certain situations. However, after all we had a number of different icons which should be visible if and only if the sunspots were in the right position.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: The views are described very accurately. If this serves an important point would an image be necessary?

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+![A bunch of icons](/img/value-of-e2e-testing/ikonit.JPG)
+
+##The First Try with Wicket Unit Tests
+
+The first automated e2e-tests were written using [Wicket Unit Tests](https://cwiki.apache.org/confluence/display/WICKET/Unit+Test). The unit tests utilized WicketTester, which runs the application without the need to actually run a web server. You can validate the web page opening, click some buttons, fill fields, run the business logic and validate the web page after. The Wicket tests are pretty quick to run and fairly simple to write. But there are also disadvantages, the tests are not actually testing the full pipe, because the lack of web server and browser. Also, the Wicket page tester does not create a picture of the page, meaning that in a case of exception or assertion error, the actual web page cannot be seen. It is like trying to find a needle in a dark room. Furthermore, the actual page structure can be seen only in debug output, i.e. finding the components to assert is not always easy.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "..does not create a picture of the page" -> render the page?

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+The first automated e2e-tests were written using [Wicket Unit Tests](https://cwiki.apache.org/confluence/display/WICKET/Unit+Test). The unit tests utilized WicketTester, which runs the application without the need to actually run a web server. You can validate the web page opening, click some buttons, fill fields, run the business logic and validate the web page after. The Wicket tests are pretty quick to run and fairly simple to write. But there are also disadvantages, the tests are not actually testing the full pipe, because the lack of web server and browser. Also, the Wicket page tester does not create a picture of the page, meaning that in a case of exception or assertion error, the actual web page cannot be seen. It is like trying to find a needle in a dark room. Furthermore, the actual page structure can be seen only in debug output, i.e. finding the components to assert is not always easy. 
+
+However, despite the impediments, these tests were found very helpful. The object was reached, the tests checked that the icons were visible and clickable in the specified situations, and even more, if you broke functionality by accident it was exposed by a test.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "the object was reached" -> should this be objective? This sentence is also very long.

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+![An example of JBehave and Selenium in action](/img/value-of-e2e-testing/jbehave.png)
+
+At first, there was a huge gap between code implemented and code tested with JBehave tests. There were a couple of developers just writing tests for the already implemented features. However, soon it was clear, that the gap would never be surpassed this way. Everyone in the project should learn writing the new type of tests. In my opinion, in the beginning it was slow and not very easy but step by step it became a natural way to write tests. Of course, the test framework was developed side by side with the tests, because we wrote a lot of common utility methods and tools.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "gap would never be surpassed" -> this does not translate well into english

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+The JBehave tests were slower to run than Wicket tests, because opening the browser took time. When the number of JBehave tests grew, we realized that running all the e2e tests on our own workstations was too time consuming. Hudson, which is our CI platform, was configured to run all the tests after a bunch of commits. This way we could run only the relevant tests on our own workstations and the full regression tests were run by Hudson. Hudson is also configured to save the print screen images if JBehave tests failed. The images are a huge help when trying to figure out why the test failed. The inspected element might be missing or maybe the test is not clicking the right element at all. The JBehave tests are also worthwhile when debugging the application. The baseline can be set up programmatically before the browser is started. After that, you can run the test step by step watching similarly the actual action on the web page.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: There are sentences both in past and present tense. Most are in past tense except "Hudson, which is.."

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+To some extent test stories can be used in communication between customers and developers or as a basis of user manual. However, soon we found out that keeping the stories in such a documentative way, is hard and on top of that not really beneficial. Also, the lack of IDE support for writing story files and matching the steps in the stories and in the test code is a little frustrating. It would be nice to find mismatches because of typos before you run the test, or navigate directly from a story step to the code implementing the step. One can say, that a number of text searches were done to avoid the problem. Furthermore, the finding right web elements to assert is not effortless with xpath either.
+
+Our end users were using Internet Explorer, but however, we could not get the tests robust enough using IE. The tests were failing quite randomly time after time. Hence, we used Firefox to run the e2e test. Even though the tests used wrong browser the benefit was obvious. Furthermore, we had problems updating the Firefox version but after we freezed the Firefox in the certain version and cut out the auto updates, everything was fine.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "cut out" -> disabled

File: _posts/2016-07-11-value-of-automated-e2e-testing.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+##The Value Obtained
+
+The comprehensive regression test suite including both the e2e tests and the unit tests gives depelopes courage to refactor the code base more freely. After all, all development is easier, because the mistakes are quite likely to expose on some layer of tests. This is extremely valuable when new developers are introduced to the project. In a long-lasting project you cannot avoid coders changing, some are coming and some are going. Of course, the rookies cannot have the all functionality and information in their heads as the experienced, tough warriors. Hence, the automated tests lower the step when entering to a new project, and as I think, ease the mental burden of both newcomers and old-timers. I would say, that extensive testing also enables healthy and needed rotation of developers between projects. The extensive automated testing is crucial also in the maintenance phase of project. The changes to code base might be done less frequently and the developers making the changes may not be the original developers any more.
```
Comment (mprencipe) [inline _posts/2016-07-11-value-of-automated-e2e-testing.md]: "coders" -> personnel

**Pull request merged**